index.d.ts

```diff
@@ -1,288 +1,620 @@
 
-  export interface Cell<D = {}> extends TableInstance<D> {
-    cell: { value: any }
-    column: Column<D>
-    row: Row<D>
-    render: (type: 'Cell' | 'Aggregated', userProps?: any) => any
-    isGrouped?: boolean
-    isAggregated?: boolean
-    isRepeatedValue?: boolean
-    getCellProps: () => any
+  export interface UseTableHeaderGroupProps<D extends object> {
+    headers: ColumnInstance<D>[]
+    getHeaderGroupProps: (props?: object) => object
+    // Added by helpers
+    totalHeaderCount: number
   }
 
-  export interface Row<D = {}> {
+  export interface UseTableColumnProps<D extends object> {
+    isVisible: boolean
+    render: (type: 'Header' | string, props?: object) => ReactNode
+    getHeaderProps: (props?: object) => object
+    // added by default
+    parent: ColumnInstance<D>
+    depth: number
     index: number
+  }
+
+  export interface UseTableRowProps<D extends object> {
     cells: Cell<D>[]
-    getRowProps: (userProps?: any) => any
-    original: D
-    path: any[]
-    values: any[]
-    depth: number
+    values: Record<IdType<D>, any>
+    getRowProps: (props?: object) => object
+    index: number
+    original: D[StringKey<D>]
+    path: IdType<D>[]
+    subRows: Row<D>[]
+    state: object
   }
 
-  export interface UseExpandedRow<D = {}> {
-    subRows?: D[]
-    groupByID?: string | number
-    toggleExpanded?: () => any
-    isExpanded?: boolean
-    isAggregated?: boolean
+  export interface UseTableCellProps<D extends object> {
+    column: ColumnInstance<D>
+    row: Row<D>
+    value: unknown
+    getCellProps: (props?: object) => object
+    render: (type: 'Cell' | string, userProps?: object) => ReactNode
   }
 
-  export type AccessorFn<D> = (
-    originalRow: D,
-    index: number,
-    sub: {
-      subRows: [D]
-      depth: number
-      data: [D]
-    }
-  ) => unknown
-
-  export interface HeaderColumn<D> {
-    /**
-     * This string/function is used to build the data model for your column.
-     */
-    accessor: IdType<D> | AccessorFn<D>
-    Header?: ReactNode | ((props: TableInstance<D>) => ReactNode)
-    Filter?: ReactNode | ((props: TableInstance<D>) => ReactNode)
-    Cell?: ReactNode | ((cell: Cell<D>) => ReactNode)
-    /**
-     * This is the unique ID for the column. It is used by reference in things like sorting, grouping, filtering etc.
-     */
+  type HeaderProps<D extends object> = TableInstance<D> & {
+    column: ColumnInstance<D>
+  } & Record<string, any>
+  type CellProps<D extends object> = TableInstance<D> & {
+    column: ColumnInstance<D>
+    row: Row<D>
+    cell: Cell<D>
+  } & Record<string, any>
+
+  // NOTE: Atleast one of (id | accessor | Header as string) required
+  export interface Accessor<D extends object> {
+    accessor?:
+      | IdType<D>
+      | ((
+          originalRow: D,
+          index: number,
+          sub: {
+            subRows: D[]
+            depth: number
+            data: D[]
+          }
+        ) => string)
     id?: IdType<D>
```

Comment:
I think that this is problematic.

When defining columns, you provide either, an id or an accessor.  If the accessor is a function then you also must provide an id.  However when the extended column definition is passed on after construction, regardless of the accessor type, there is an id provided.  So outside of the initial column definition, the id shouldn't be optional.

Reply:
perhaps add `id: IdType<D>` to the ColumnInstance ?

Reply:
You're correct. I modified the `ColumnInstance` to not use the definition of `id` in `Column`